Remove `cwd` as a CommandOption and replace with the new `config.root…

februarycat opened the pull request:
## Changes

- Removes the option to use `cwd` as a CommandOption
- Add new config option `root` (defaults to `process.cwd()`) so a user can use it to specify the root of a project using Snowpack either programmatically or as a config option

## Testing

- Updated most of the build tests
- Refer to [this comment](https://github.com/snowpackjs/snowpack/pull/1928/#issuecomment-744505648) for a extensive breakdown of the tests changes

## Docs

- The docs are yet to be updated. Will work on that after double-checking the refactor is complete and new tests are in place

drwpow commented on snowpack/src/config.ts:
```diff
@@ -40,6 +40,7 @@ const ALWAYS_EXCLUDE = ['**/node_modules/**/*', '**/web_modules/**/*', '**/.type
 
 // default settings
 const DEFAULT_CONFIG: SnowpackUserConfig = {
+  root: process.cwd(),
```
This looks good! This looks like the correct default behavior to me.

I think in addition to this line, we also need to add resolution below in the `loadConfigurationForCLI()` method when we’re loading a cosmiconfig. For example, say a user has their project set up like so:

```
├── config/
│   └── snowpack.config.js
└── src/
```

And say in their config they set:

```
{
  root: '../src',
}
```

**We should resolve this from the location of `config/snowpack.config.js`.** As is, it’s unclear how `src` will resolve. And since we do want to eventually follow this behavior, treating your config file location as root (similar to TSConfig, etc.), I think it’s better to do it in the same PR where we introduce the option. That way we won’t break setups for people if they start using this now, and we change that behavior later.

februarycat replied:
I already started working on `loadConfigurationForCLI()` but I got a little stuck because I wasn't sure whether to add the `root` as a flag to the CLI too. I'll work on it the next days and update the MR to get your opinion.

FredKSchott replied:
@drwpow I'd actually like to tackle that problem in a separate PR, if you don't mind. It's a bigger ask than I think you realize. We have lots of other properties in the config that also need this support that are currently always relative to `process.cwd()` (`mount`, `plugins`, `alias`, all file paths like `out`, etc.). 

We also need to decide if we continue to support `extends`. If we do, we'll also need to handle this support relative to the extended config. But, I think since we've begun leaning more on JS config than JSON lately, I think we could also just skip this work and remove `extends` support in v3 in favor just whatever you want to script in JS.

Last reason: it means that most existing tests will continue to work as-is in this PR. Given that this touches so many different parts of the codebase, test stability is a plus at the moment.

We can still keep thrash down by only merging this into the v3 release branch.

FredKSchott commented:
@februarycat this is great! We definitely want to get this in soon to test for v3, so I hope you don't mind that I went ahead and updated the other `process.cwd()` references across the codebase to now use `root`. 

Would still love your help on some tests and otherwise getting this cleaned up for review!

FredKSchott commented:
@februarycat this is great! We definitely want to get this in soon to test for v3, so I hope you don't mind that I went ahead and updated the other `process.cwd()` references across the codebase to now use `root`. 

As mentioned in the comment above, there are a few remaining references to `process.cwd()`, mostly around config file loading from the CLI and how we normalize user-given relative config values into absolute paths. Those can remain as-is for now, to be resolved in a follow-up PR. For now, assume all relative URLs in config continue to resolve relative to `process.cwd()`.

~~Would still love your help on some tests and otherwise getting this cleaned up for review!~~ Actually, I just realized that the best way to test this may be to start moving some tests over from CLI tests to JS API tests (this is one of the big things we've been looking forward to once `root` support arrives). No tests that use root will pass if `root` isn't working correctly. 

Check out the test that I just converted: ce198a6c. It would be great if you could grab a few more tests and convert them to use the JS API this way. Just getting ~3 done in this PR would give confidence enough to merge then, and then follow up with the remaining tests in follow up PRs.

februarycat commented:
@FredKSchott Sure, I will update the tests following your example. If I have time I'll try to fiddle with the remaining `process.cwd()` to see if I can make it work. If not, we can merge just the updated tests.

februarycat commented:
Fix messed up merge!

februarycat commented:
Okay, so I've updated most of the tests in `/test/build` and fixed some typos in `/plugins`. Here's a breakdown of my changes:

`/test/build`:
Updated most of the tests to use the JS API (create a snowpack config and build programmatically) as in [ce198a6](https://github.com/snowpackjs/snowpack/commit/ce198a6c196981e7d6eb3d9d61dfc0971d5e3213) example.
The updates are:
- Update test.js file
- Remove config file or config from package.json and move that in the test file
- Remove `testbuild` script (following the example commit from @FredKSchott, although I recommend keeping some default command for standalone usage)

There are some tests that still need updating but they rely on `extends`, `proxy`, `plugin` functionality which was not updated with cwd -> root functionality yet:

- `config-alias` (left untouched as it depends on a relative path plugin and the functionality wasn't added)
- `config-extends-plugin` (left untouched as it tests `extends` functionality which, as I understood it, it might be deprecated and the `root` functionality wasn't added)
- `config-instantiated-object` (contains relative plugin)
- `config-out-flag` (tests CLI flag)
- `config-ts-file` (tests .ts config which already defaults to `root: process.cwd()`)
- `css-modules` (build is correct but and all files are generated but .proxy are not returned from `snowpack.buildProject()`, might have to do with the root/cwd updates which are not complete)
- `html-environment-variables` (tests environment variables from the command line)
- `import-assets` (no proxy files in `snowpack.buildProject()` result)
- `import-json` (no proxy files in `snowpack.buildProject()` result)
- `plugin-hook-optimize` (contains relative plugin path)
- `plugin-hook-transform` (contains relative plugin path)

I updated `package-bootstrap` as the tests passes because it's not checking the existence of `.proxy` files in the result from `snowpack.buildProject()`. It expects to find `_dist_/index.js` containing an import to the proxy, which is correctly generated.

Some of the above tests could be upgraded but I decided to skip them as I wasn't sure if they needed a refactor. If any of these could benefit from being updated to use the JS API let me know and I'll update them!

`/plugins`:
- `plugin-run-script` test was failing because the plugin was instantiated without a config object. I fixed it by passing an empty object so it defaults to `process.cwd()`
- `plugin-typescript` was failing for the same reason as the tests above. I passed `{}` as config so the tests pass
- - `plugin-typescript/plugin.js` renamed `snowpackConfig_` to `snowpackConfig` to keep naming consistent across all tests
- `web-test-runner-plugin` was failing after @FredKSchott did the initial refactor then decided to revert some changes. The test was failing because `cwd` was not defined. Fixed it by replacing `cwd` with `snowpackConfig.root || process.cwd()` but this needs more checking from @FredKSchott

FredKSchott commented:
Huge! This is awesome, will take a look today! 

If this PR didn't touch every plugin, I'd say go for it. But, because it's already so sprawling I'd like to keep it scoped as is for now. You don't need to update every test in this PR, especially if you're interested in doing this work in a followup with the remaining tests.

mark as ready for review, and I'll give this a look 👍

februarycat commented:
Marked as ready for review!

I'd love to keep working on this. I've recently found out about snowpack and instantly loved it. I want to keep contributing 💯

FredKSchott commented on snowpack/src/config.ts:
```diff
@@ -638,6 +642,7 @@ function normalizeConfig(_config: SnowpackUserConfig): SnowpackConfig {
       '`devOptions.out` is now `buildOptions.out`! `devOptions.out` will be deprecated in the next major release.',
     );
   }
+  config.root = path.resolve(cwd, config.root);
```
I believe one of your failing tests is because `installOptions.cwd` should be set to root:

```
config.installOptions = config.installOptions || {};
config.installOptions.cwd = config.root;
```

This is needed because `installOptions` are passed to esinstall, which is a seperate tool that needs to be told about the new root. It's okay that we keep using `installOptions.cwd` for now.

februarycat replied:
Fixed in [a8530aa](https://github.com/snowpackjs/snowpack/pull/1928/commits/a8530aa68c317697d2b2e6536b09d5ddcd3c1c23)

FredKSchott commented on test/build/base-url-homepage/base-url-homepage.test.js:
```diff
@@ -1,26 +1,40 @@
+let result;
 
-let files = {};
+function getFile(id) {
```
We should move this into `test-utils` so that we don't need to define it individually in every test file. Maybe `getBuildResult` instead since `getFile` feels a bit "Node.js fs library"

februarycat replied:
Would this be okay?

```javascript
getBuildResult(result, TEST_OUT, './index.html');
```

The function has to take in the `result` object and the `TEST_OUT` param as they might be different for each test. Or maybe I should curry the function to take the first 2 params on beforeAll?

FredKSchott commented on test/build/base-url-homepage/base-url-homepage.test.js:
```diff
@@ -1,26 +1,40 @@
+    const [, config] = snowpack.createConfiguration({
+      root: TEST_ROOT,
+      mount: {
+        [path.resolve(TEST_ROOT, './public')]: '/',
```
+1 for creating a fresh config object in every test like you do here, and not refactoring this setup into some helper. It's more clear what each test is testing imo.

februarycat replied:
Would this be okay?

```javascript
getBuildResult(result, TEST_OUT, './index.html');
```

The function has to take in the `result` object and the `TEST_OUT` param as they might be different for each test. Or maybe I should curry the function to take the first 2 params on beforeAll?

februarycat replied:
Tbh I though of doing it but then I had the same thought, it's easier to understand what's happening when you see everything in the test file. A helper would've looked more complicated + you'd still have to pass down lots of params for every test to config it which would result in even more unnecessary abstraction.

februarycat replied:
I could create a helper eventually if it becomes useful to have one in the future.

FredKSchott commented on test/build/legacy-mount-scripts/legacy-mount.test.js:
```diff
@@ -1,22 +1,33 @@
 describe('config: mount scripts (legacy)', () => {
-  beforeAll(() => {
-    setupBuildTest(__dirname);
+  beforeAll(async () => {
```
I'm actually not sure why this test is failing, but because this support will be going out in v3.0 and we're removing this legacy `mount` support in v3.0 anyway, I'm +1 for just removing this test of legacy behavior.

februarycat replied:
Resolved in (73bdab6)[https://github.com/snowpackjs/snowpack/pull/1928/commits/73bdab6de514f3452138f66d5e781601af4e4beb]

FredKSchott reviewed (commented):
Looks good! I think I gave hints on the two failing tests, let me know if those get your CI passing and then we should be good to merge! 

/cc @drwpow @matthewp since this also touches (build) test code

FredKSchott commented:
LGTM! I think the test failure is unrelated (or if it is related, it's impossible to tell what's going on from CI). Going to merge this and then will get that last test passing on our end. 

Thanks for the change! Looking forward to getting that "everything is relative to cwd instead of the file itself" issue resolved as well!

FredKSchott commented:
hmm, it seems like there's something more serious going on in the tests. I couldn't reproduce on my local, but I also couldn't get CI running either. 

I went ahead and rolled back out all of the testing changes from `main`, for now. It would be good to go back and add only 1-3 to confirm that they work on CI long-term, before converting all of them. We also are unlocked to create the automatic fixture test-runner, which could be a better target to convert to (no need to do two separate conversions if we can just jump straight to the final version)

februarycat commented:
Should I go ahead and open a new PR with base-* tests updated? I think there's only three of them which as you suggested should be enough for now.

FredKSchott commented:
Yes, that sounds perfect. I think @drwpow wanted a chance to review the test changes as well so that should be a great PR on its own
